src/csvcubed/writers/qbwriter.py

```diff
@@ -284,18 +311,123 @@ def _generate_foreign_keys_for_cube(self) -> List[dict]:
 
         return foreign_keys
 
-    def _generate_virtual_columns_for_cube(self) -> List[Dict[str, Any]]:
+    def _generate_virtual_columns_for_obs_val_in_pivoted_shape_cube(
+        self,
+    ) -> List[Dict[str, Any]]:
+        virtual_columns: List[dict] = []
+
+        # Generate the virtual col defining the `?sliceUri rdf:type qb:Slice` triple.
+        virtual_columns.append(
+            {
+                "name": "virt_slice",
+                "virtual": True,
+                "propertyUrl": "rdf:type",
+                "valueUrl": "qb:Slice",
+            }
+        )
+
+        virtual_columns.append(
+            {
+                "name": "virt_slice_structure",
+                "virtual": True,
+                "propertyUrl": "qb:sliceStructure",
+                "valueUrl": self._new_uri_helper.get_slice_key_across_measures_uri(),
+            }
+        )
+
+        # For each obs val column returned by dsd_type function
+        observation_value_columns = get_columns_of_dsd_type(
+            self.cube, QbObservationValue
+        )
+        for obs_column in observation_value_columns:
+            # Creating a virtual col for `?sliceUri qb:observation ?obsUri`
+            observation_uri = self._get_observation_uri_for_pivoted_shape_data_set(
```

Comment:
This function is getting a bit big now, could you refactor the stuff in the obs_column for loop out into a separate function?